File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/abilities/AbilityGraph.java
package de.limbusdev.guardianmonsters.guardians.abilities;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ArrayMap;

import de.limbusdev.guardianmonsters.guardians.items.equipment.BodyPart;
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian;
import de.limbusdev.guardianmonsters.guardians.monsters.SpeciesDescription;

/**
 * @author Georg Eckert 2017
 */

public class AbilityGraph implements IAbilityGraph
{
    private AGuardian core;

    private ArrayMap<Integer, Node> nodes;
    private Array<Edge> edges;

    private ArrayMap<Integer,Ability.aID> abilityNodes;
    private ArrayMap<Integer,BodyPart> equipmentNodes;
    private Array<Integer> metamorphosisNodes;

    private ArrayMap<Integer, Ability.aID> activeAbilities; // Abilities available in battle
    private ArrayMap<Integer, Ability.aID> learntAbilities; // Abilities activated on the graph
    private Array<BodyPart> learntEquipment;

    public AbilityGraph(AGuardian core, SpeciesDescription data)
    {
        nodes = new ArrayMap<>();
        edges = new Array<>();
        abilityNodes = new ArrayMap<>();
        equipmentNodes = new ArrayMap<>();
        learntAbilities = new ArrayMap<>();
        learntEquipment = new Array<>();
        metamorphosisNodes = new Array<>();

        for(int i = 0; i < GraphTemplate.coords.length; i++) {
            int v[] = GraphTemplate.coords[i];
            nodes.put(i, new Node(v[X+1], v[Y+1], i));
        }

        for(int i = 0; i < GraphTemplate.conns.length; i++) {
            int e[] = GraphTemplate.conns[i];
            edges.add(new Edge(nodes.get(e[X]), nodes.get(e[Y])));
        }

        activeAbilities = new ArrayMap<>();
        for(int i=0; i<7; i++) {
            activeAbilities.put(i,null);
        }

        init(core, data);
    }

    public AbilityGraph(
            ArrayMap<Integer, Node> nodes,
            Array<Edge> edges,
            ArrayMap<Integer, Ability.aID> abilityNodes,
            ArrayMap<Integer, BodyPart> equipmentNodes,
            Array<Integer> metamorphosisNodes,
            ArrayMap<Integer, Ability.aID> activeAbilities,
            ArrayMap<Integer, Ability.aID> learntAbilities,
            Array<BodyPart> learntEquipment)
    {
        this.nodes = nodes;
        this.edges = edges;
        this.abilityNodes = abilityNodes;
        this.equipmentNodes = equipmentNodes;
        this.metamorphosisNodes = metamorphosisNodes;
        this.activeAbilities = activeAbilities;
        this.learntAbilities = learntAbilities;
        this.learntEquipment = learntEquipment;
    }

    private void init(AGuardian core, SpeciesDescription data)
    {
        this.core = core;

        for(int key : data.getAbilityNodes().keys())
        {
            nodes.get(key).type = Node.Type.ABILITY;
            abilityNodes.put(key, data.getAbilityNodes().get(key));
        }

        for(int key : data.getEquipmentNodes().keys())
        {
            nodes.get(key).type = Node.Type.EQUIPMENT;
            equipmentNodes.put(key, data.getEquipmentNodes().get(key));
        }

        for(int key : data.getMetamorphosisNodes())
        {
            nodes.get(key).type = Node.Type.METAMORPHOSIS;
            metamorphosisNodes.add(key);
        }

        int counter = 0;
        for(Ability.aID a : learntAbilities.values())
        {
            activeAbilities.put(counter, a);
            counter++;
        }

        activateNode(0);
    }


    // ........................................................................... GETTERS & SETTERS
    @Override
    public ArrayMap<Integer, Node> getNodes() {
        return nodes;
    }

    @Override
    public Array<Edge> getEdges() {
        return edges;
    }

    @Override
    public ArrayMap<Integer, Ability.aID> getAbilityNodes()
    {
        return abilityNodes;
    }

    @Override
    public ArrayMap<Integer, BodyPart> getEquipmentNodes()
    {
        return equipmentNodes;
    }

    @Override
    public Array<Integer> getMetamorphosisNodes()
    {
        return metamorphosisNodes;
    }

    @Override
    public void activateNode(int nodeID)
    {
        Node node = nodes.get(nodeID);

        node.activate();
        enableNeighborNodes(nodeID);

        switch(node.type) {
            case ABILITY:
                learnAbility(nodeID);
                break;
            case EQUIPMENT:
                learnEquipment(nodeID);
                break;
            default:
                break;
        }

        core.setAbilitiesChanged();
        core.notifyObservers();
    }

    @Override
    public boolean isNodeEnabled(int nodeID) {
        return nodes.get(nodeID).isEnabled();
    }

    @Override
    public boolean learnsAbilityAt(int nodeID) {
        return abilityNodes.containsKey(nodeID);
    }

    @Override
    public boolean learnsEquipmentAt(int nodeID) {
        return equipmentNodes.containsKey(nodeID);
    }

    public boolean metamorphsAt(int nodeID) {
        return metamorphosisNodes.contains(nodeID, false);
    }

    @Override
    public int getCurrentForm()
    {
        int activatedMetamorphosisNodes = 0;
        for(int key : metamorphosisNodes)
        {
            if(isNodeEnabled(key)) activatedMetamorphosisNodes++;
        }
        return activatedMetamorphosisNodes;
    }

    @Override
    public boolean learnsSomethingAt(int nodeID)
    {
        return (learnsAbilityAt(nodeID) || learnsEquipmentAt(nodeID) || metamorphsAt(nodeID));
    }

    @Override
    public Node.Type nodeTypeAt(int nodeID) {
        if(learnsAbilityAt(nodeID)) {
            return Node.Type.ABILITY;
        }
        if(learnsEquipmentAt(nodeID)) {
            return Node.Type.EQUIPMENT;
        }
        if(metamorphsAt(nodeID)) {
            return Node.Type.METAMORPHOSIS;
        }
        return Node.Type.EMPTY;
    }

    @Override
    public boolean hasLearntEquipment(BodyPart bodyPart) {
        if(learntEquipment.contains(bodyPart,true)) {
            return true;
        } else {
            return false;
        }
    }

    @Override
    public Ability.aID getActiveAbility(int abilitySlot)
    {
        return activeAbilities.get(abilitySlot);
    }

    @Override
    public Ability.aID getRandomActiveAbility()
    {
        return getActiveAbility(MathUtils.random(0, activeAbilities.size-1));
    }

    @Override
    public void setActiveAbility(int slot, int learntAbilityNumber)
    {
        Ability.aID abilityToLearn = learntAbilities.get(learntAbilityNumber);
        if(abilityToLearn == null) return;

        for(int key : activeAbilities.keys()) {
            Ability.aID abilityAtThisSlot = activeAbilities.get(key);

            if(abilityAtThisSlot != null) {
                if (abilityAtThisSlot.equals(abilityToLearn)) {
                    activeAbilities.put(key, null);
                }
            }
        }
        activeAbilities.put(slot, learntAbilities.get(learntAbilityNumber));

        core.setAbilitiesChanged();
        core.notifyObservers();
    }

    @Override
    public ArrayMap<Integer, Ability.aID> getActiveAbilities() {
        return activeAbilities;
    }


    // .............................................................................. HELPER METHODS
    private boolean learnAbility(int nodeID)
    {
        if(learnsAbilityAt(nodeID)) {
            learntAbilities.put(nodeID, abilityNodes.get(nodeID));
            return true;
        } else {
            return false;
        }
    }

    private boolean learnEquipment(int nodeID) {
        if(learnsEquipmentAt(nodeID)) {
            learntEquipment.add(equipmentNodes.get(nodeID));
            return true;
        } else {
            return false;
        }
    }

    private void enableNeighborNodes(int nodeID) {
        for (Edge e : edges) {
            if (e.from.ID == nodeID || e.to.ID == nodeID) {
                int enableID = e.from.ID == nodeID ? e.to.ID : e.from.ID;
                nodes.get(enableID).enable();
            }
        }
    }
}

File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/abilities/AbilityGraph.kt
package de.limbusdev.guardianmonsters.guardians.abilities

import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ArrayMap

import de.limbusdev.guardianmonsters.guardians.items.equipment.BodyPart
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian
import de.limbusdev.guardianmonsters.guardians.monsters.SpeciesDescription
import de.limbusdev.utils.extensions.set

/**
 * @author Georg Eckert 2019
 */

class AbilityGraph : IAbilityGraph
{
    // --------------------------------------------------------------------------------------------- PROPERTIES

    // .............................................................. public
    // Interface Properties
    override val nodes              get() = _nodes
    override val edges              get() = _edges
    override val abilityNodes       get() = _abilityNodes
    override val equipmentNodes     get() = _equipmentNodes
    override val metamorphosisNodes get() = _metamorphosisNodes
    override val activeAbilities    get() = _activeAbilities
    override val learntAbilities    get() = _learntAbilities
    override val learntEquipment    get() = _learntEquipment

    // .............................................................. private

    // Backing Property Fields
    private var _nodes              : ArrayMap<Int, Node>         = ArrayMap()
    private var _edges              : Array<Edge>                 = Array()
    private var _abilityNodes       : ArrayMap<Int, Ability.aID>  = ArrayMap()
    private var _equipmentNodes     : ArrayMap<Int, BodyPart>     = ArrayMap()
    private var _metamorphosisNodes : Array<Int>                  = Array()
    private var _activeAbilities    : ArrayMap<Int, Ability.aID?> = ArrayMap()
    private var _learntAbilities    : ArrayMap<Int, Ability.aID>  = ArrayMap()
    private var _learntEquipment    : Array<BodyPart>             = Array()

    private lateinit var core       : AGuardian // Guardian this graph belongs to


    // --------------------------------------------------------------------------------------------- CONSTRUCTORS
    constructor
    (
            nodes               : ArrayMap<Int, Node>,
            edges               : Array<Edge>,
            abilityNodes        : ArrayMap<Int, Ability.aID>,
            equipmentNodes      : ArrayMap<Int, BodyPart>,
            metamorphosisNodes  : Array<Int>,
            activeAbilities     : ArrayMap<Int, Ability.aID?>,
            learntAbilities     : ArrayMap<Int, Ability.aID>,
            learntEquipment     : Array<BodyPart>
     ){
        _nodes              = nodes
        _edges              = edges
        _abilityNodes       = abilityNodes
        _equipmentNodes     = equipmentNodes
        _metamorphosisNodes = metamorphosisNodes
        _activeAbilities    = activeAbilities
        _learntAbilities    = learntAbilities
        _learntEquipment    = learntEquipment
    }

    constructor(core: AGuardian, data: SpeciesDescription)
    {
        // Assemble ability graph layout: Nodes
        for (i in GraphTemplate.coordinates.indices)
        {
            val nodePosition = GraphTemplate.getNodePosition(i)
            nodes.put(i, Node(nodePosition.x, nodePosition.y, i))
        }

        // Assemble ability graph layout: Edges
        for (i in GraphTemplate.connections.indices)
        {
            val edge = GraphTemplate.getEdge(i)
            edges.add(Edge(nodes[edge.x], nodes[edge.y]))
        }

        initialize(core, data)
    }



    private fun initialize(core: AGuardian, data: SpeciesDescription)
    {
        this.core = core

        for (key in data.abilityNodes.keys())
        {
            nodes[key].type = Node.Type.ABILITY
            abilityNodes[key] = data.abilityNodes[key]
        }

        for (key in data.equipmentNodes.keys())
        {
            nodes[key].type = Node.Type.EQUIPMENT
            equipmentNodes[key] = data.equipmentNodes[key]
        }

        for (key in data.metamorphosisNodes)
        {
            nodes[key].type = Node.Type.METAMORPHOSIS
            metamorphosisNodes.add(key)
        }

        var counter = 0
        for(slot in 0..6) { activeAbilities[slot] = null }
        learntAbilities.values().forEach { a -> activeAbilities[counter] = a; counter++ }

        activateNode(0)
    }


    // --------------------------------------------------------------------------------------------- INTERFACE

    // .............................................................. Interface Properties
    override val currentForm: Int get()
    {
        var activeMetamorphosisNodeCounter = 0
        for (key in metamorphosisNodes)
        {
            if (nodes[key].isActive) { activeMetamorphosisNodeCounter++ }
        }
        return activeMetamorphosisNodeCounter
    }


    // .............................................................. Interface Methods
    override fun activateNode(nodeID: Int)
    {
        val node = nodes.get(nodeID)

        node.activate()
        enableNeighborNodes(nodeID)

        when (node.type)
        {
            Node.Type.ABILITY   -> learnAbility(nodeID)
            Node.Type.EQUIPMENT -> learnEquipment(nodeID)
            else                -> {}
        }

        core.setAbilitiesChanged()
        core.notifyObservers()
    }

    override fun setActiveAbility(slot: Int, learntAbilityNumber: Int)
    {
        check(slot in 0..6)
        check(learntAbilities.containsKey(learntAbilityNumber))
        val newActiveAbility = learntAbilities[learntAbilityNumber]
        checkNotNull(newActiveAbility)

        // Check if chosen ability already occupies a slot, -1 if not present
        val oldSlotOfNewAbility = activeAbilities.indexOfValue(newActiveAbility, false)

        // If Ability has already been active, clear the previous slot
        if(oldSlotOfNewAbility != -1) { activeAbilities[oldSlotOfNewAbility] = null }

        // Get ability that is currently in this slot
        val oldActiveAbility = activeAbilities[slot]

        // If slot was preoccupied, reset it
        if(oldActiveAbility != null) { activeAbilities[slot] = null }

        // Put new ability into the slot
        activeAbilities[slot] = newActiveAbility

        // If the new slot was already occupied and the new ability has been active, swap slots
        if(oldSlotOfNewAbility != -1 && oldActiveAbility != null)
        {
            activeAbilities[oldSlotOfNewAbility] = oldActiveAbility
        }

        core.setAbilitiesChanged()
        core.notifyObservers()
    }

    override fun isNodeEnabled(nodeID: Int) = nodes[nodeID].isEnabled

    override fun learnsAbilityAt(nodeID: Int) = abilityNodes.containsKey(nodeID)

    override fun learnsEquipmentAt(nodeID: Int) = equipmentNodes.containsKey(nodeID)

    override fun metamorphsAt(nodeID: Int) = metamorphosisNodes.contains(nodeID, false)

    override fun learnsSomethingAt(nodeID: Int): Boolean
    {
        return learnsAbilityAt(nodeID) || learnsEquipmentAt(nodeID) || metamorphsAt(nodeID)
    }

    override fun nodeTypeAt(nodeID: Int): Node.Type
    {
        return when
        {
            learnsAbilityAt(nodeID)   -> Node.Type.ABILITY
            learnsEquipmentAt(nodeID) -> Node.Type.EQUIPMENT
            metamorphsAt(nodeID)      -> Node.Type.METAMORPHOSIS
            else                      -> Node.Type.EMPTY
        }
    }

    override fun hasLearntEquipment(bodyPart: BodyPart): Boolean
    {
        return learntEquipment.contains(bodyPart, true)
    }

    override fun getRandomActiveAbility() : Ability.aID
    {
        val slot = getRandomActiveAbilitySlot()
        val ability = activeAbilities[slot]

        checkNotNull(ability) { "Random ability is null. The must always be at least one active ability." }
        return ability
    }

    override fun getRandomActiveAbilitySlot() : Int
    {
        return MathUtils.random(0, activeAbilities.size - 1)
    }


    // --------------------------------------------------------------------------------------------- HELPER METHODS
    private fun learnAbility(nodeID: Int): Boolean
    {
        if(learnsAbilityAt(nodeID))
        {
            learntAbilities.put(nodeID, abilityNodes.get(nodeID))
            return true
        }
        return false
    }

    private fun learnEquipment(nodeID: Int): Boolean
    {
        if (learnsEquipmentAt(nodeID))
        {
            learntEquipment.add(equipmentNodes.get(nodeID))
            return true
        }
        return false
    }

    /** Usually only neighbors of active nodes may be activated as well. This marks them as ENABLED. */
    private fun enableNeighborNodes(nodeID: Int)
    {
        for (e in edges)
        {
            if (e.from.ID == nodeID || e.to.ID == nodeID)
            {
                val enableID = if (e.from.ID == nodeID) e.to.ID else e.from.ID
                nodes.get(enableID).enable()
            }
        }
    }
}
